Change behavior of apply_dimension and reduce

m-mohr opened the pull request:
Based on issue #73: Made apply_dimension more usable and made reduce more restricted to real "reduce" use cases.

- `apply_dimension` was basically replaced with a completely new definition.
- `reduce`: Reducers can't return multiple values any longer. Parameter `target_dimension` was therefore removed.

What do you think?

soxofaan commented on apply_dimension.json:
```diff
@@ -37,7 +38,14 @@
+            "required": true
+        },
+        "target_dimension": {
+            "description": "The name of the target dimension or `null` to replace the dimension specified in `dimension` (default).\n\nIf a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original type.\n\nThe resulting target di ...
```
proposal to simplify things a bit

> The new name of the dimension that replaces the original dimension, or `null` (the default) to reuse the original name and type of the dimension. 
>
> The resulting target dimension has as many dimension labels ...

m-mohr replied:
Hmm, yes, there's room for improvement. 
First, the first sentence is better capturing my intention as follows:
> The name of the target dimension or `null` to replace the dimension labels and pixel values in the original dimension (default)

I didn't really intend a drop&add (= replace) of the dimension. If that's desired, specify a target_dimension.

What I'm also thinking about is sorting, for example. It's just the values that change, but also reference system etc should stay intact. What I'm no wondering about is: Should the dimension labels be preserved when sorting? What about preserving the labels when computing the cumsum, for example? This is not really captured any longer with the "enumerated" labels. Hmm...

soxofaan replied:
ok, I indeed was wondering about the intended precedence of that sentence "name or null to replace ...":
- (name or null) to replace
- name or (null to replace)

I assumed the first, but you meant the second

soxofaan replied:
If I understand correctly, there are quite some aspects you want to control through the `target_dimension` argument:
- dimension name (reuse original or rename)
- dimension type (reuse original or "other")
- dimension labels (reuse originals, new integer range, something else?)

I guess it's a bit much to try to encode this in a single string/null type argument.
Maybe `target_dimension` should be an object with dedicated (but optional) fields, to control each of these aspects?

m-mohr replied:
Thanks for the feedback. I tried to make it more clear:

> The name of the target dimension or `null` (default).
> 
> 1. The default behavior is to replace the dimension labels and pixel values in the original dimension.
> 2. If a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original type.
> 
> Afterwards, the dimension has as many dimension labels as the array returned by the process has values. For example, a process returning an array with four elements results in a dimension with four labels.

m-mohr replied:
> I guess it's a bit much to try to encode this in a single string/null type argument.
> Maybe `target_dimension` should be an object with dedicated (but optional) fields, to control each of these aspects?

No, if you want to control these settings, use add_dimension beforehand:
> If a dimension with the specified name exists, the dimension is replaced, but keeps the original settings.

I clarified the target_dimension description again and hope it covers all cases now (see also https://github.com/Open-EO/openeo-processes/pull/111#issuecomment-567473532):
> The name of the target dimension or `null` (default).
>
> 1. The default behavior is to replace the pixel values (and dimension labels if the number of pixel values and dimension labels doesn't match) in the original dimension.
> 2. If a target dimension is specified, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created. If a dimension with the specified name exists, the dimension is replaced, but keeps the original name, type, reference system (and dimension labels if the number of values matches the number of dimension labels).
> 
>Afterwards, the dimension has as many dimension labels as the array returned by the process has values. For example, a process returning an array with four elements results in a dimension with four labels.

Does that make more sense?

soxofaan replied:
> If a target dimension is specified, a new dimension ... is created

Maybe I'm reading this wrong, but this allows the interpretation that the old dimension/data is preserved (the new dimension is just added), which practically and mathematically makes no sense in the general case. 

> If a dimension with the specified name exists, the dimension is replaced ....

I don't get this. How can you replace another dimension than the one specified by `dimension` argument? For example you have cube with dims "foo" and "bar", you do apply_dimension on "foo", and you use "bar" as target_dimension, that doesn't make sense to me.

m-mohr replied:
You are right, the source dimension would have to be removed. I'll clarify that...

m-mohr replied:
Tried to update the description for target_dimension:
> The name of the target dimension or `null` (the default) to use the source dimension specified in the parameter `dimension`.
>
> * The default is that the target dimension is the source dimension so that the pixel values are replaced in the source dimension. The original name, type and reference system are preserved.
> * By specifying a target dimension, the source dimension is removed and the target dimension is populated with the computed pixel values.
>     * If the target dimension *does exist*, a new dimension with the specified name and the type `other` (see ``add_dimension()``) is created.
>     * If the target dimension *does not exist*, the pixel values are replaced in the target dimension. The original name, type and reference system are preserved.
>
> In general, the target dimension will have as many dimension labels as values are available after applying the process. The dimension labels are preserved when the number of pixel values in the source dimension is equal to the number of values computed by the process. Otherwise, the dimension labels will be incrementing integers starting from zero.

Does it make more sense now?

soxofaan replied:
Yes, that's better. 

But overall it seems like a very long explanation for what is basically a name change. 
For example, isn't it possible to merge the part "if the target dim does exist ..." into the part about the default  behaviour? The case where the target dimension does exist but is different from the source dim should be illegal anyway, right?

m-mohr replied:
You are really pushing me into short and crisp writing, @soxofaan ;-)

Do you like commit https://github.com/Open-EO/openeo-processes/pull/111/commits/806bb7dde8bcd8df606d5c0f7b6f3881cca48752 better? I hope I still cover all cases good enough...

soxofaan commented on reduce.json:
```diff
@@ -83,7 +72,7 @@
     },
     "returns": {
-        "description": "A data cube with the newly computed values. The number of dimensions is reduced by one, except if a target dimension is specified. The resolution and cardinality are the same as for the original data cube.",
+        "description": "A data cube with the newly computed values. The number of dimensions is reduced by one. The resolution and cardinality are the same as for the original data cube.",
```
> The number of dimensions is reduced by one. The resolution and cardinality are the same as for the original data cube."

"It has one dimension less than the original data cube, but the remaining dimensions have their resolution and cardinality preserved."

m-mohr replied:
Sounds good, fixed.

m-mohr commented on apply.json:
```diff
@@ -1,7 +1,7 @@
-    "summary": "Applies a unary process to each pixel",
-    "description": "Applies a **unary** process which takes a single value such as `abs` or `sqrt` to each pixel value in the data cube (i.e. a local operation). In contrast, the process ``apply_dimension()`` applies an n-ary process to a particular dimension.",
+    "summary": "Applies a process to each pixel",
+    "description": "Applies a *unary* process to all pixel values in the data cube (i.e. a local operation). A unary process takes a single value and returns a single value, for example ``abs()`` or ``linear_scale_range()``. In contrast, the process ``apply_dimension()`` applies a process to all pixels along a particular dimension.",
```
Indeed, fixed.

m-mohr commented on apply_dimension.json:
```diff
@@ -1,14 +1,15 @@
-    "summary": "Applies an n-ary process to all pixels",
-    "description": "Applies an **n-ary** process (i.e. takes an array of pixel values instead of a single pixel value) to a raster data cube. In contrast, the process ``apply()`` applies an unary process to all pixel values.\n\nBy default, `apply_dimension` applies the process on all pixel values in the data cube as ``apply()`` does, but the parameter `dimension` can be specified to work only on  ...
+    "summary": "Applies a process to pixels along a dimension",
+    "description": "Applies a process to all pixel values along a dimension of a raster data cube. For example, if the temporal dimension is specified the process will work on a time series of pixel values.\n\nBy default, the dimension is replaced with a dimension of the same type. The dimension will result in as many enumerated dimension labels as array elements are available afterwards. Enumera ...
```
> > The dimension will result in as many enumerated dimension labels as array elements are available afterwards.
> 
> I'm not sure I'm understanding this. Maybe you mean "the dimension will have .."?

Yes, indeed.

> > enumerated dimension labels
> 
> Is it necessary to use "enumerated" here? Explicit use of it sounds to me like it is the responsibility of the user to provide a enumeration of the labels.

I just ported that over from reduce. The thing I want to make clear here is that the labels afterwards have numerated dimension labels.

> > process `apply()` applies a process to all pixel values ...
> 
> "to each pixel individually"

Thanks, fixed.

m-mohr commented:
What I'm also thinking about is sorting, for example. It's just the values that change, but also reference system etc should stay intact. What I'm no wondering about is: Should the dimension labels be preserved when sorting? What about preserving the labels when computing the cumsum, for example? This is not really captured any longer with the "enumerated" labels.

mkadunc commented:
[Re @m-mohr](https://github.com/Open-EO/openeo-processes/pull/111#issuecomment-567473532)
> What I'm also thinking about is sorting, for example. It's just the values that change, but also reference system etc should stay intact. What I'm no wondering about is: Should the dimension labels be preserved when sorting? What about preserving the labels when computing the cumsum, for example? This is not really captured any longer with the "enumerated" labels.

When sorting values, the quantity (variable) remains the same, but the labels change - you go from e.g. time axis `time{t_1, t_2, .., t_N}` to rank axis `rank{1,2,..,N}`.

When computing cumsum, the quantity changes (you go from value-at-t_i to sum-value-between-t_0-and-t_i); the labels stay the same, but they (might) change in meaning:
- before cumsum labels on the `time` axis are points-in-time when the measurement happended;
- after cumsum, labels effectively denote the right boundary of the time interval for which the values apply; left boundary of that same interval is the first of the original labels

mkadunc commented on apply_dimension.json:
```diff
@@ -48,16 +56,13 @@
     },
     "returns": {
-        "description": "A data cube with the newly computed values for the specified . The resolution, cardinality and the number of dimensions are the same as for the original data cube.",
+        "description": "A data cube with the newly computed values for the specified. The resolution, cardinality and the number of dimensions are the same as for the original data cube.",
```
Cardinality may change

```suggestion
        "description": "A data cube with the newly computed values for the specified. The resolution and the number of dimensions are the same as for the original data cube.",
```

m-mohr commented:
> When sorting values, the quantity (variable) remains the same, but the labels change - you go from e.g. time axis `time{t_1, t_2, .., t_N}` to rank axis `rank{1,2,..,N}`.

Ah, okay. So sort/order don't work well with apply_dimension now:
>  The dimension labels are preserved when the number of pixel values in the source dimension is equal to the number of values computed by the process.

There's more work needed, also having in mind what @soxofaan wrote in https://github.com/Open-EO/openeo-processes/pull/111#discussion_r363208494 . Will give it a try....

m-mohr commented:
Telco: Agreed, but rename reduce to reduce_dimension.

m-mohr commented:
Huh, so we also rename filter to filer_dimension?!

m-mohr commented:
Updated the PR, I think we can merge with one or two reviews.

I also renamed filter to filter_labels as it seems that's what is the direction also with reduce_dimension.

Later, I need to check whether there are references to reduce and filter in API or the docs (e.g. the glossary) and update them.

m-mohr commented:
To not further block my work, merging without approvals. If there's still something to change, open a new PR for it.
